Cognite.Common/TaskThrottler.cs

```diff
@@ -281,7 +296,23 @@ private async Task Run()
                 {
                     if (_timeUnit > TimeSpan.Zero)
                     {
-                        WaitHandle.WaitAny(new[] { token.WaitHandle, _taskCompletionEvent }, _timeUnit);
+                        // If waiting is interrupted, we want to continue waiting, instead of restarting,
+                        // this helps ensure even distribution of tasks. I.e. if we run tasks that takes 200ms,
+                        // and use a time unit of 500ms, then we will wake up after 200ms, not schedule a new task,
+                        // wait another 500ms, then wake up again, so we effectively only wake up every 700ms,
+                        // This wake-up period should be a minimum value.
+                        TimeSpan waitTime = _timeUnitChunk;
+                        var now = DateTime.UtcNow;
+                        if (_lastWaitTime + _timeUnitChunk > now)
+                        {
+                            waitTime = _timeUnitChunk - (now - _lastWaitTime);
+                        }
+
+                        if (_lastWaitTime == DateTime.MinValue) _lastWaitTime = DateTime.UtcNow;
+
+                        var res = WaitHandle.WaitAny(new[] { token.WaitHandle, _taskCompletionEvent }, waitTime);
```

Comment:
This method blocks the calling thread... would it make any sense to put it inside `await Task.Run()` and send it to other thread?

Reply:
Good point. I think there might be a way to do this asynchronously properly, using Task.Run isn't quite right here I think.